### PR TITLE
Autodetect/install/update iotedgehubdev

### PR DESCRIPTION
**Pull request by adashen**
1. During extension activation, it will check whether iotedgehubdev needs to be upgraded/installed. And it is not a force option. Note: if pip not detected, nothing will happen.
2. When trigger iotedgehubdev related command, it will check whether iotedgehubdev is installed, it is a force option. And corresponding error will be shown to user. And it will not upgrade the iotedgehubdev if installed.

Please help to have a look at the wording whether it deliver the message correctly. 
And telemetry related stuff will be added in later PRs

### CODE REVIEW

File: src/common/constants.ts [hunk 1]
```diff
@@ -142,6 +142,16 @@ export class Constants {
     public static amlWorkspaceDesc = "Azure Machine Learning Workspace";
     public static asaJobDesc = "Azure Stream Analytics Job";
     public static amlApiVersion: string = "2018-03-01-preview";
+
+    public static needSimulatorInstalledMsg = "You must have the 'iotedgehubdev' tool installed for Edge Hub simulation.";
```
Comment (LazarusX) [inline src/common/constants.ts]: `Edge Hub simulation` -> `IoT Edge simulator` for alignment and avoiding introducing new concept?

File: src/common/constants.ts [hunk 2]
```diff
@@ -142,6 +142,16 @@ export class Constants {
     public static amlWorkspaceDesc = "Azure Machine Learning Workspace";
     public static asaJobDesc = "Azure Stream Analytics Job";
     public static amlApiVersion: string = "2018-03-01-preview";
+
+    public static needSimulatorInstalledMsg = "You must have the 'iotedgehubdev' tool installed for Edge Hub simulation.";
+    public static updateSimulatorMsg = "Update your 'iotedgehubdev' tool to the latest for the best experience.";
+    public static failedInstallSimulator = "Failed to install 'iotedgehubdev' tool because of error:";
+    public static outputNoSimulatorMsg = "Cannot execute command since there is no 'iotedgehubdev' installed. Please install it first.";
```
Comment (LazarusX) [inline src/common/constants.ts]: `there is no 'iotedgehubdev' installed` -> `'iotedgehubdev' is not installed`

File: src/common/constants.ts [hunk 3]
```diff
@@ -142,6 +142,16 @@ export class Constants {
     public static amlWorkspaceDesc = "Azure Machine Learning Workspace";
     public static asaJobDesc = "Azure Stream Analytics Job";
     public static amlApiVersion: string = "2018-03-01-preview";
+
+    public static needSimulatorInstalledMsg = "You must have the 'iotedgehubdev' tool installed for Edge Hub simulation.";
+    public static updateSimulatorMsg = "Update your 'iotedgehubdev' tool to the latest for the best experience.";
+    public static failedInstallSimulator = "Failed to install 'iotedgehubdev' tool because of error:";
+    public static outputNoSimulatorMsg = "Cannot execute command since there is no 'iotedgehubdev' installed. Please install it first.";
+    public static pipNotFoundMsg = "'pip' not found. Cannot install 'iotedgehubdev' tool and will have to be installed manually.";
+    public static installFailedMsg = "'iotedgehubdev' tool installation has been failed and  will have to be installed manually.";
```
Comment (LazarusX) [inline src/common/constants.ts]: Grammatical fix: `has been failed` -> `has failed`

Comment (LazarusX) [inline src/common/constants.ts]: `and  will ` -> `and will ` (extra space between)

File: src/common/constants.ts [hunk 4]
```diff
@@ -142,6 +142,16 @@ export class Constants {
     public static amlWorkspaceDesc = "Azure Machine Learning Workspace";
     public static asaJobDesc = "Azure Stream Analytics Job";
     public static amlApiVersion: string = "2018-03-01-preview";
+
+    public static needSimulatorInstalledMsg = "You must have the 'iotedgehubdev' tool installed for Edge Hub simulation.";
+    public static updateSimulatorMsg = "Update your 'iotedgehubdev' tool to the latest for the best experience.";
+    public static failedInstallSimulator = "Failed to install 'iotedgehubdev' tool because of error:";
+    public static outputNoSimulatorMsg = "Cannot execute command since there is no 'iotedgehubdev' installed. Please install it first.";
+    public static pipNotFoundMsg = "'pip' not found. Cannot install 'iotedgehubdev' tool and will have to be installed manually.";
```
Comment (LazarusX) [inline src/common/constants.ts]: ` will have to be installed manually` -> `please install it manually` to clarify that it is the user himself/herself who needs to perform that action.

File: src/edge/simulator.ts [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            let input: vscode.MessageItem | undefined;
+            if (force) {
+                input = await vscode.window.showWarningMessage(message, { modal: true }, ...items);
+            } else {
+                input = await vscode.window.showWarningMessage(message, ...items);
+            }
+
+            if (input === install) {
+                try {
+                    await Executor.executeCMD(outputChannel, "pip", {shell: true}, "install --upgrade iotedgehubdev");
```
Comment (LazarusX) [inline src/edge/simulator.ts]: This may fail on *nix because of permission issue.

File: src/edge/simulator.ts [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    Failed,
+    Canceled,
+    NoPip,
+}
+
+export class Simulator {
+    public static async validateSimulatorUpdated(outputChannel: vscode.OutputChannel = null): Promise<void> {
+        let message: string;
+        try {
+            const output = await Executor.executeCMD(undefined, "iotedgehubdev", { shell: true }, "--version");
```
Comment (LazarusX) [inline src/edge/simulator.ts]: The iotedgehubdev command here (and below) runs in output channel, so users' virtual env won't work.

File: src/extension.ts [hunk 7]
```diff
@@ -191,16 +194,27 @@ function initCommandAsync(context: vscode.ExtensionContext,
...
                 properties.result = "Failed";
-                errorData = new ErrorData(error);
-                outputChannel.appendLine(`Error: ${errorData.message}`);
-                vscode.window.showErrorMessage(errorData.message);
+                if (error instanceof LearnMoreError) {
+                    const learnMore: vscode.MessageItem = { title: Constants.learnMore };
+                    const items: vscode.MessageItem[] = [ learnMore ];
+                    outputChannel.appendLine(`Error: ${error.message}`);
+                    if (await vscode.window.showErrorMessage(error.message, ...items) === learnMore) {
+                        await opn(error.url);
```
Comment (LazarusX) [inline src/extension.ts]: How about `vscode.open`? I saw that the Java debugger extension just switched from `opn` to `vscode.open` for unknown reasons: https://github.com/Microsoft/vscode-java-debug/blob/master/CHANGELOG.md#changed

File: src/edge/simulator.ts (hunk 5 above)
Comment (adashen) [reply src/edge/simulator.ts]: pip install should not ask for sudo right? but run it needs sudo.

File: src/edge/simulator.ts (hunk 6 above)
Comment (adashen) [reply src/edge/simulator.ts]: I would like fix virtual env as an issue. Does that mean virtual env is not setup in output? only for terminal?

File: src/edge/simulator.ts (hunk 5 above)
Comment (LazarusX) [reply src/edge/simulator.ts]: Here is how vscode-python installs Python packages: https://github.com/Microsoft/vscode-python/blob/master/src/client/common/installer/moduleInstaller.ts#L36-L40

Comment (LazarusX) [reply src/edge/simulator.ts]: If users have `python.pythonPath` set, the `iotedgehubdev start` command which stills runs in terminal, will pick up users `python.pythonPath` setting. Installing iotedgehubdev with output channel will have no effect.

File: src/edge/simulator.ts (hunk 6 above)
Comment (LazarusX) [reply src/edge/simulator.ts]: Virtualenv only applies to terminals.

File: src/edge/simulator.ts (hunk 5 above)
Comment (LazarusX) [reply src/edge/simulator.ts]: @adashen Whether `pip install` requires `sudo` really depends how users installs Python. For example, on macOS, if the user uses the built-in Python, `sudo` is required, but not for Homebrewed Python. This is the same for Linux.

Comment (adashen) [reply src/edge/simulator.ts]: will check the platform, for windows install for user, else pop message and ask user to install manually.

File: src/edge/simulator.ts (hunk 6 above)
Comment (adashen) [reply src/edge/simulator.ts]: will update install page to ask user install iotedgehubdev in global python env. New feature request may be specify the iotedgehubdev path of virtual env.

File: src/extension.ts (hunk 7 above)
Comment (adashen) [reply src/extension.ts]: Thanks, will update

File: src/edge/simulator.ts [hunk 8]
```diff
@@ -79,6 +85,10 @@ export class Simulator {
     }
 
     private static async installSimulatorWithPip(force: boolean, message: string, outputChannel: vscode.OutputChannel = null): Promise<InstallReturn> {
+        // auto install only supported on windows. For linux/max ask user install manually.
```
Comment (LazarusX) [inline src/edge/simulator.ts]: `max` -> `macOS`

**Pull request merged**